broker.go

```diff
@@ -135,7 +142,8 @@ func (b *Broker) RegisterNode(id NodeID, node Node) error {
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
-	b.nodes[id] = node
+	b.nodes[id] = nodeRegistration{node: node, usages: 0}
```

Comment:
should we prevent nodes from being registered (overwritten) in the broker?

Reply:
It's a good point, I was just following the existing behaviour when I modified this.

The way it works at the moment it feels idempotent. If it prevented re-registration/overwriting when the ID already exists, on subsequent calls would we:

* return an error 
* do nothing and return`nil` as we do now. 
 
Returning `nil` and no error feels like it might mislead the caller that the node they tried to register was indeed registered, but it would just be the old node still kicking about.

I wonder are the pros/cons, is there a use case where you'd want to swap out an existing node for a new node 'on the fly'?  What kind of problems could we run into if we did change a node when we didn't mean to 🙈 

🤔

Reply:
Oh, I just noticed ... we can still allow this, but we should maintain existing counts/references/usages.

Reply:
Yes, we should keep the same counts/usages!   Perhaps we need an option (so it's not a breaking change) which optionally requests an error if the node exists?  Right now, I don't think there's a way for callers to even know if someone else has registered a node with the same ID.

Reply:
> Right now, I don't think there's a way for callers to even know if someone else has registered a node with the same ID.

Yes that's right, but we never checked if we were overwriting a node. 

> Perhaps we need an option (so it's not a breaking change) which optionally requests an error if the node exists?

I'm drawn towards leaving the method signature as it is currently, updating Go Doc comments and returning an an error if you try to register a node with an ID is already in use.

Leaving the method signature 'as-is' does mean existing consumers might not realise that we've changed the underlying behaviour, as their code still compiles. 😞 

It feels like the original intent was likely not to allow overwriting as that would likely have some odd side effects for pipelines which were already using that node. I do wonder if anyone would be making use of the existing behaviour today or if they'd consider it a bug. 🤔

Reply:
Or potentially have `NewBroker` accept `...options` for example, to specify whether nodes can be overwritten? 🤔 Just thinking out loud.